vscode-dotnet-runtime-library/src/Utils/NodeIPCMutex.ts

```diff
@@ -0,0 +1,286 @@
+ * https://github.com/microsoft/vscode/blob/main/extensions/git/src/ipc/ipcServer.ts#L15
+ * https://github.com/microsoft/vscode/blob/main/src/vs/code/electron-main/main.ts#L318
+ */
+export class NodeIPCMutex
+{
+    private readonly lockPath: string;
+    private server?: Server;
+    private hasCleanedUpBefore = false;
+    /**
+     *
+     * @param lockId - The ID of the lock. This should be a unique identifier for the lock being created.
+     * The lockId must be less than about 90 characters as it will get truncated down to 107 characters -- if it is longer than the temp folder may allow on OS X, it may fail.
+     * // https://nodejs.org/api/net.html#:~:text=other%20operating%20systems.-,Identifying%20paths%20for%20IPC%20connections,-%23
+     * @param logger - A custom logger for your own logging events. By default, this will log to the console.
+     */
+    constructor(lockId: string, readonly logger: INodeIPCMutexLogger = new INodeIPCMutexLogger())
+    {
+        this.lockPath = this.getIPCHandlePath(lockId);
+    }
+
+    private getIPCHandlePath(id: string): string
+    {
+        const lengthLimit = os.platform() === 'win32' ? 256 : 107;
+
+        // On Unix, A file descriptor in /temp/ is a good option to hold this sock.
+        // In Linux, The user or system may set XDG_RUNTIME_DIR to set our applications temporary directory, so this this instead of /temp/
+        // On Unix, Access to /tmp/ may be restricted, but all processes must use the same directory, so we can't condition to use another dir based on the permissions for it.
+
+        // On Windows, '\\\\.\\pipe\\` is a Special File System to get a Named Pipe (File Descriptors won't work)
+        // https://nodejs.org/docs/latest/api/net.html#ipc-support:~:text=On%20Windows%2C%20the,owning%20process%20exits.
+
+        const ipcPathDir = os.platform() === 'win32' ? `\\\\.\\pipe\\` :
+            os.platform() === 'linux' && process.env.XDG_RUNTIME_DIR ? process.env.XDG_RUNTIME_DIR as string : os.tmpdir();
+
+        if (id.length > (lengthLimit - ipcPathDir.length))
+        {
+            this.logger.log(`Lock ID is too long, truncating to ${lengthLimit - ipcPathDir.length} characters, due to dir: ${ipcPathDir}`);
+            id = id.substring(0, Math.max(lengthLimit - 1, 6)); // Prevent setting the environment variable to make the application fail. the vscd prefix + 2 should uniquely identify most locks.
+        }
+
+        if (process.platform === 'win32')
+        {
+            return `${ipcPathDir}vscd-${id}-sock`;
+        }
+        else
+        {
+            return path.join(ipcPathDir, `vscd-${id}.sock`);
+        }
+    }
+
+    /**
+     * @remarks This function will try to hold a lock to prevent both other processes and other async code in this processes from running simultaneously..
```

Comment:
nit:
```suggestion
     * @remarks This function will try to hold a lock to prevent both other processes and other async code in this processes from running simultaneously.
```